### PR TITLE
Fix GPU fillholes to respect hipass thresholds below 0.5

### PR DESCRIPTION
**Pull request by evanscho**
GPU `fillholes_iterative_hipass_fill_m` hardcoded a single `> 0.5` threshold, silently dropping all mask values ≤ 0.5. The CPU version iterates 14 threshold levels.
The GPU version now also iterates down to 0.1 like the CPU version does.

```python
# Before (buggy):
inv_mask = 1.0 - (samples > 0.5).float()  # ignores everything ≤ 0.5

# After:
for threshold in [1, 0.99, 0.97, 0.95, 0.93, 0.9, 0.8, 0.7, 0.6, 0.5, 0.4, 0.3, 0.2, 0.1]:
    thresholded = (result >= threshold).float()
    if thresholded.sum() == 0:
        continue
    # binary close (max_pool2d dilation + negated erosion) → flood-fill → update
```

### GPU fix (`inpaint_cropandstitch.py`)

- Iterate all 14 thresholds with binary closing + flood-fill per level, matching CPU behavior
- Skip optimization: early-exit for thresholds with no matching pixels

### Benchmark (`test_fillholes_benchmark.py`)

- `--device {cpu,cuda}` flag (auto-detects by default); CPU scipy reference always runs on CPU
- CUDA synchronization for accurate GPU timing
- Correctness tests: pure-white (exact match), sub-0.5 ring (regression), multi-threshold (documents expected diffs)

### Performance (RTX 5090)

| Mask | CPU | Old GPU | New GPU | vs CPU | vs Old GPU |
|------|-----|---------|---------|--------|------------|
| pure-white 512×512 | 41.37 ms | 0.15 ms | 1.59 ms | **96% faster** | 960% slower |
| multi-threshold 64×64 | 1.37 ms | 0.74 ms | 5.74 ms | 319% slower | 676% slower |

The common case (pure-white masks) is still substantially faster than CPU. Multi-threshold small masks are slower due to CUDA kernel launch overhead across 14 sequential flood-fill passes.

### Why CPU/GPU results can still differ

- **Binary closing border handling**: scipy `border_value=1` vs `max_pool2d` zero-padding
- **Flood-fill connectivity**: scipy uses 4-connectivity; `max_pool2d(3×3)` uses 8-connectivity
- **Float precision**: CPU float64 (numpy) vs GPU float32

Pure-white masks match exactly. Multi-threshold: 1.6% pixels differ, max abs diff = 0.1.

### Further notes

May want to clarify to user when to prefer CPU vs GPU mode, but I'll leave that to the project maintainer to decide.

### CODE REVIEW

**Comment by lquesada** [comment]
thanks for identifying this issue.

I have some concerns with the code, though:

            for _ in range(max(H, W)):
                next_outside = TF.max_pool2d(curr_outside.unsqueeze(1), kernel_size=3, stride=1, padding=1).squeeze(1)
                next_outside = next_outside * padded_inv
                next_outside[:, 0, :] = 1
                next_outside[:, -1, :] = 1
                next_outside[:, :, 0] = 1
                next_outside[:, :, -1] = 1 
                if torch.all(next_outside == curr_outside):
                    break
                curr_outside = next_outside




The CPU implementation iterates once per image in the batch.

This implementation iterates once per max(W, H), and breaks only if torch.all(next_outside == curr_outside).

Why do we need that iteration? Also, torch.all(next_outside == curr_outside) on floats is subject to floating point offs and may actually force the algorithm to iterate max(W, H) times.

**Comment by evanscho** [comment]
Good catch on torch.all(next_outside == curr_outside) for floats! We can change to do the flood-fill in uint8 (it's a binary algorithm anyway), use bitwise & instead of float multiply, and torch.equal():

            inv_mask = (closed == 0).byte()

            padded_inv = torch.zeros((B, H+2, W+2), device=device, dtype=torch.uint8)
            padded_inv[:, 1:-1, 1:-1] = inv_mask

            curr_outside = torch.zeros((B, H+2, W+2), device=device, dtype=torch.uint8)
            curr_outside[:, 0, :] = 1
            curr_outside[:, -1, :] = 1
            curr_outside[:, :, 0] = 1
            curr_outside[:, :, -1] = 1

            for _ in range(max(H, W)):
                next_outside = TF.max_pool2d(
                    curr_outside.float().unsqueeze(1), kernel_size=3, stride=1, padding=1
                ).squeeze(1).byte()
                next_outside = next_outside & padded_inv
                next_outside[:, 0, :] = 1
                next_outside[:, -1, :] = 1
                next_outside[:, :, 0] = 1
                next_outside[:, :, -1] = 1
                if torch.equal(next_outside, curr_outside):
                    break
                curr_outside = next_outside

            filled = 1.0 - curr_outside[:, 1:-1, 1:-1].float()
            result = torch.max(result, filled * threshold)


Regarding the loop: The GPU fillholes is reimplementing scipy's binary_fill_holes (which has no GPU equivalent) by doing a flood-fill from the image border. It grows the "outside" region one pixel per iteration using 3×3 max_pool. max(H, W) is the worst case for how far the flood needs to travel. The CPU version has the same O(max(H,W)) cost — it's just hidden inside scipy's C code. The GPU version looks different because it processes the entire batch in parallel rather than looping per image.

**Comment by evanscho** [comment]
On another note: you may decide architecturally to default to always revert to CPU mode for more complex masks, and thus keep the GPU mode as is without supporting lower thresholds. This may make more sense given that CPU is more efficient for the complex cases where parallelization doesn't help.

**Comment by lquesada** [comment]
I think I may just default to CPU for this task, as it is a priori not straightforward to determine whether a mask is "too complex". I'd rather have this be less flaky and risky at the expense of some performance. Filling mask holes is in any case not the top bottleneck. WDYT?

**Comment by evanscho** [comment]
I think you're right to default to CPU, but I'd change the comment in parentheses to say something like CPU (best for masks with blurs) and GPU (best for solid white masks).

And if you want to be fancy about it you can have GPU mode revert to CPU automatically when it's processing masks with thresholds 0.5 or lower.

I think in either case, it makes sense to revert the other pull request too.

**Comment by lquesada** [comment]
not sure if I managed to explain myself correctly but I just pushed it. Both CPU and GPU will run the CPU logic, so it will work with all masks the same.
